qa/tasks/deepsea.py

```diff
@@ -363,8 +404,64 @@ def _introspect_roles(self, deepsea_ctx):
         for var in report_vars:
             self.log.info("{} == {}".format(var, deepsea_ctx[var]))
 
+    def _master_python_version(self, py_version):
+        """
+        Determine if a given python version is installed on the Salt Master
+        node.
+        """
+        python_binary = 'python{}'.format(py_version)
+        installed = True
+        try:
+            self.master_remote.run(args=[
+                'type',
+                python_binary,
+                run.Raw('>'),
+                '/dev/null',
+                run.Raw('2>&1'),
+                ])
+        except CommandFailedError:
+            installed = False
+        if installed:
+            self.master_remote.run(args=[
+                python_binary,
+                '--version'
+                ])
+        else:
+            self.log.info(
+                '{} not installed on master node'.format(python_binary)
+                )
+        return installed
+
+    def _maybe_apply_alternative_defaults(self):
+        global_yml = '/srv/pillar/ceph/stack/global.yml'
+        if self.alternative_defaults:
+            data = ''
+            for cnf in self.alternative_defaults:
+                for k, v in cnf.items():
+                    info_msg = (
+                        "Applying alternative default {}: {}".format(k, v)
+                        )
+                    data += "{}: {}\n".format(k, v)
```

Comment:
This code is bad, since you work with yaml output, you better use yaml api (like yaml.dump) for generating such kind of data. If self.alternative_defaults list items contain dictionaries with duplicating keys will the generated config be valid for you? If so, then following oneliner should generate valid yaml code.
```
data = "\n".join([yaml.dump(x, default_flow_style=False) 
                  for x in self.alternative_defaults])
```
I don't see any advantage of accumulating "Applying..." messages since the whole data can be logged. Like:

```
self.master_remote.sh('sudo tee -a %s' % global_yml, input=data')
```

Reply:
cool, added to wip-qa-cleanup

(I'm going to merge these four PRs as-is, since they have been exhautively tested, and make any changes in follow-up PRs.)

Reply:
> cool, added to wip-qa-cleanup
> 
> (I'm going to merge these four PRs as-is, since they have been exhautively tested, and make any changes in follow-up PRs.)
this is a good explanation for me when I would like to merge the code without review :-DDD